doc: add acrn-sanity-test (#3684)
projectacrn/acrn-hypervisor: doc/getting-started/acrn-sanity-test.rst

ACRN Sanity Test
################

The document describes the daily sanity test for ACRN by validation team, which 
includes a set of system level core tests of basic functionality to demonstrate 
ACRN hypervisor’s availability. User could manually run these after setup ACRN. 

> 

Reviewer: I have just realised that you use the pre-installed script (`launch_uos.sh`) in these tests. It assumes the system has a user called `clear` already set-up. We need to make this explicit in this document as most people will set-up a different user.

Reviewer: Are you going to add some text here to address this comment or will @deb-intel do it?

* Test Steps

1. Boot DUT, and enter SOS from ACRN hypervisor

   .. code-block:: none

      $ vm_console

#. Check IP and bridge/tap after log in sos

   .. code-block:: none
     
      $ ip a
      $ brctl show
      bridge name     bridge id               STP enabled     interfaces
      acrn-br0         8000.1ad4ac453b20       no              acrn_tap0
                                                               ens4

#. Boot UOS using the launch_uos.sh 

   .. code-block:: none
   
      $ bash /usr/share/acrn/samples/nuc/launch_uos.sh

#. Check UOS IP address

* Expected Result

IP address should be allocated for both SOS and UOS, and ping is ok

ST_ACRNIO_LaaG_Virtio_storage
*****************************

* Description

This test case focus on checking if LaaG using virtio storage can boot up successfully.

* Prerequisites

Need support Virtio

* Test Steps

1. In SOS: copy or create one img file as rootfs.
#. Boot UOS using this img file in cmd line in launch_uos.sh

   .. code-block:: none

   $ cd /home/clear/uos/

   $ wget https://download.clearlinux.org/releases/xxxxx/clear/clear-xxxxx-kvm.img.xz

>    $ unxz clear-xxxxx-kvm.img.xz

Reviewer: ```suggestion
   $ unxz uos.img.xz
```

>    $ vim launch_uos.sh

Reviewer: And now you can remove this step altogether.

   -s 3,virtio-blk,/home/clear/uos/uos.img \
   
   to

   -s 3,virtio-blk,/home/clear/uos/xxxxx.img \

* Expected Result

UOS can boot up successfully

ST_ACRNIO_SOS_Ethernet
**********************

* Description

This test case focus on checking if IP can be allocated to Service OS correctly.

* Prerequisites

Network card is installed

* Test Steps

1. Boot into SOS

> #. IP a in SOS

Reviewer: Do you mean that the user should run `ip a` from a terminal in the SOS?

* Expected Result

#. IP address (either local or external) should be allocated for sos(1 ip address), and ping is ok.

ST_ACRNIO_SOS_Storage
*********************

* Description

> This test case focus on checking if EMMC can work normally in Service OS.

Reviewer: ```suggestion
This test case focus on checking if eMMC works correctly in Service OS.
```

* Prerequisites

EMMC is installed

* Test Steps

1. Boot into SOS

> #. fdisk -l in SOS

Reviewer: We need to turn these into a `code-block` section.

> #. Find EMMC block and mount it to read/write

Reviewer: ```suggestion
#. Find eMMC block and mount it to read/write
```

Please also change all other occurences of "EMMC" to "eMMC"

   NVME disk in /dev/mmcblk1p3
   SATA disk in /dev/sda3

* Expected Result

EMMC can be used normally in SOS

> GVT

Reviewer: Please put a more explicit title, e.g. **Graphics Virtualization (GVT)**

***

ST_ACRNGT_SOS_Rendering_Weston_APP
**********************************

* DescriptioACRNCRNGT provide virtual display on LaaG: run weston /weston-simple-egl in LaaG,
workload will show normally on LaaG

* Prerequisites

System boot up normally

Install ias weston related bundle

   .. code-block:: none

>       $ swupd bundle-add software-defined-cockpit x11-server --allow-insecure-http

Reviewer: You need to run `swupd` as root (with `sudo`).

Also curious why you need to use the `--allow-insecure-http` flag?

> Please note: do not install bundle desktop, these two may had conflict

Reviewer: Please make this a proper `note::` section

* Test Steps

1. Boot up hypervisor and SOS
#. Start ias weston by command in SOS&LaaG:

   .. code-block:: none

>      $ systemctl enable ias

Reviewer: You need to be `sudo` for all these `systemctl` operations

> ACRN Sanity Test

Reviewer: Are these test cases described below valid for *all* HW platforms we support? If not, let's call out which are expected to work as described.

################

The document describes the daily sanity test for ACRN by validation team, which 
includes a set of system level core tests of basic functionality to demonstrate 
ACRN hypervisor’s availability. User could manually run these after setup ACRN. 

ACRN Sanity Test Cases
######################

Hypervisor
##########

Introduction
************

ST_HV_Bootloader_LaaG_VUEFI_Default
**********************************

* Description

> The test case focus on testing boot up a LaaG with VUEFI and default launch_uos.sh.

Reviewer: ```suggestion
The test case focus on testing boot up a LaaG using a virtual UEFI (OVMF) firmware and the default ``launch_uos.sh``.
```

* Prerequisites

Service OS boots up normally

* Test Steps

1. Launch SOS
#. Use ``acrn-dm`` to launch clearlinux using the pre-installed default ``launch_uos.sh`` script

  .. code-block:: none

     $ sudo /usr/share/acrn/samples/nuc/launch_uos.sh

* Expected Result

LaaG boot up normally

> ST_HV_CPU_LaaG_Partition_Check

Reviewer: Is this test case still relevant? The number of CPU cores assigned to a VM is no longer handled by `acrn-dm` (via the `launch_uos.sh` script)?

******************************

* Description

This test case focus on checking if the pass through CPU to Guest OS is 
identical as ``launch_uos.sh`` defined.

* Prerequisites

Service OS and Guest OS boot up normally

* Test Steps

1. Boot SOS (Default all CPUs will be seen under SOS)
#. Check CPU core by ``cat /proc/cpuinfo``
#. Boot LaaG UOS with default launch_uos.sh (pass through 1 core to UOS)
#. Check CPU status by ``cat /proc/cpuinfo``

* Expected Result

1. Step 2: Expect to see all CPUs under SOS
#. Step 4: Expect to see 1 core under UOS since launch_uos.sh pass-through 1 core

ST_HV_Memory_LaaG_Default_Partition_Check
*****************************************

* Description

This test case focus on checking if the pass through memory to Guest OS and 
check if the memory size is in a reasonable scope as launch_uos.sh defined. 
Since the Guest OS will take a part of memory, the available memory size will 
not be exact as what pass-through.

* Prerequisites

Service OS and Guest OS boot up normally

* Test Steps

1. Boot to SOS
#. Boot LaaG UOS with default memory setting in launch_uos.sh

> #. Check ‘cat /proc/meminfo’ and see if the available memory is close the size passed through to UOS

Reviewer: ```suggestion
#. Check ``cat /proc/meminfo`` and see if the available memory is close the size passed through to UOS
```

   .. code-block:: none

   $ mkdir -p /home/clear/uos/

   $ cd /home/clear/uos/

   $ curl https://download.clearlinux.org/releases/xxxxx/clear/clear-xxxxx-kvm.img.xz -o uos.img.xz

   $ unxz uos.img.xz

* Expected Result

UOS can boot up successfully

ST_ACRNIO_SOS_Ethernet
**********************

* Description

This test case focus on checking if IP can be allocated to Service OS correctly.

* Prerequisites

Network card is installed

* Test Steps

1. Boot into SOS  
   
   .. code-block:: none

   $ ip a

* Expected Result

#. IP address (either local or external) should be allocated for sos(1 ip address), and ping is ok.

ST_ACRNIO_SOS_Storage
*********************

* Description

This test case focus on checking if eMMC works correctly in Service OS.

* Prerequisites

eMMC is installed

* Test Steps

1. Boot into SOS

   .. code-block:: none
  
   $ fdisk -l 

#. Find eMMC block and mount it to read/write
   NVME disk in /dev/mmcblk1p3
   SATA disk in /dev/sda3

* Expected Result

eMMC can be used normally in SOS

Graphics Virtualization (GVT)
*****************************

ST_ACRNGT_SOS_Rendering_Weston_APP
**********************************

> * DescriptioACRNCRNGT provide virtual display on LaaG: run weston /weston-simple-egl in LaaG,

Reviewer: Can you fix the text/title on this line?

TC_ACRNGT_SOS_LaaG_Rendering_WestonAPP
**************************************

* Description

ACRNGT support GPU sharing feature, run weston apps weston-simple-shm in SOS and LaaG,
both SOS and UOS will show workload normally

* Prerequisites

System boot up normally
Install ias weston related bundle

   .. code-block:: none

      $ sudo swupd bundle-add software-defined-cockpit x11-server 

   .. note::

   do not install bundle desktop, these two may had conflict

* Test Steps

1. Boot up hypervisor, SOS and LaaG
#. Start ias weston by command in SOS&LaaG:

   .. code-block:: none

      $ sudo systemctl enable ias
      $ sudo systemctl start ias
      $ sudo export XDG_RUNTIME_DIR=/run/ias

#. cd /usr/share/ias/examples/
#. Run ./weston-simple-shm in SOS & LaaG

* Expected Result

Able to run test apps in SOS and LaaG synchronously

ST_ACRNGT_LaaG_Rendering_Weston_OpenGL
**************************************

* Description

ACRNGT provide virtual display on LaaG: run weston /weston-simple-egl in LaaG,
workload will show normally on LaaG

* Prerequisites

System boot up normally Install ias weston related bundle

   .. code-block:: none

>       $ swupd bundle-add software-defined-cockpit x11-server --allow-insecure-http

Reviewer: Do you really need the `--allow-insecure-http` flag here?

* Test Steps

1. Boot up hypervisor, SOS and LaaG
#. Start ias weston by command in LaaG:

   .. code-block:: none

>       $ systemctl enable ias

Reviewer: Need `sudo`

      $ systemctl start ias
      $ export XDG_RUNTIME_DIR=/run/ias

#. cd /usr/share/ias/examples/
#. Run ./weston-simple-egl in LaaG

* Expected Result

Able to run rotating colorful triangle in the screen

   .. code-block:: none

>        $ weston-simple-egl

Reviewer: Move this `code-block::` section above and merge with the two bullets

* Prerequisites

Need support Virtio

* Test Steps

1. In SOS: copy or create one img file as rootfs.
#. Boot UOS using this img file in cmd line in launch_uos.sh

   .. code-block:: none

   $ cd /home/clear/uos/

   $ wget https://download.clearlinux.org/releases/xxxxx/clear/clear-xxxxx-kvm.img.xz

   $ unxz clear-xxxxx-kvm.img.xz

   $ vim launch_uos.sh

   -s 3,virtio-blk,/home/clear/uos/uos.img \
   
   to

   -s 3,virtio-blk,/home/clear/uos/xxxxx.img \

* Expected Result

UOS can boot up successfully

ST_ACRNIO_SOS_Ethernet
**********************

* Description

This test case focus on checking if IP can be allocated to Service OS correctly.

* Prerequisites

Network card is installed

* Test Steps

1. Boot into SOS
#. IP a in SOS

* Expected Result

#. IP address (either local or external) should be allocated for sos(1 ip address), and ping is ok.

ST_ACRNIO_SOS_Storage
*********************

* Description

This test case focus on checking if EMMC can work normally in Service OS.

* Prerequisites

EMMC is installed

* Test Steps

1. Boot into SOS
#. fdisk -l in SOS
#. Find EMMC block and mount it to read/write

>    NVME disk or mrb  in /dev/mmcblk1p3

Reviewer 2: mrb should not open source.
